### PR TITLE
Mtdsa 4105

### PR DESCRIPTION
**Pull request by BetaDraconis**

### CODE REVIEW

File: app/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetail.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+object EoyEstimateDetail extends NestedJsonReads {
+
+  val emptyEoyEstimateDetail: EoyEstimateDetail = EoyEstimateDetail()
+
+  implicit val writes: OWrites[EoyEstimateDetail] = Json.format[EoyEstimateDetail]
+  implicit val reads: Reads[EoyEstimateDetail] = (
+    readSequenceAndMapToType[EoyEstimateSelfEmployment](JsPath \ "incomeSource") and
+      readSequenceAndMapToType[EoyEstimateUkPropertyFHL](JsPath \ "incomeSource", "04").map(_.getOrElse(Seq.empty).headOption) and
+      readSequenceAndMapToType[EoyEstimateUkPropertyNonFHL](JsPath \ "incomeSource", "02").map(_.getOrElse(Seq.empty).headOption) and
```
Comment (jeremystone) [inline app/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetail.scala]: If I deliberately break the code and swap (e.g.) the "02" and the "04" you have for the FHL non-FHL, the tests still pass.

Maybe you should use a different value for the taxable income in order to differential the Json object for the different income source types.

File: test/v1/fixtures/endOfYearEstimate/detail/EoyEstimateSelfEmploymentFixtures.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+package v1.fixtures.endOfYearEstimate.detail
+
+import play.api.libs.json.{ JsObject, JsValue, Json }
+import v1.models.response.getCalculation.endOfYearEstimate.detail.EoyEstimateSelfEmployment
+
+object EoyEstimateSelfEmploymentFixtures {
+
+  val selfEmploymentID: String   = "AA123456789"
+  val taxableIncome: BigInt      = 1011
```
Comment (jeremystone) [inline test/v1/fixtures/endOfYearEstimate/detail/EoyEstimateSelfEmploymentFixtures.scala]: (See above comment) Use different values for each piece of Json and the corresponding models in these fixtures.

(If there were a string field you could use it to more clearly identify them.)

File: test/v1/models/response/getCalculation/GetCalculationResponseSpec.scala [hunk 3]
```diff
@@ -124,10 +168,75 @@ class GetCalculationResponseSpec extends UnitSpec {
...
+      |		  "ukPropertyFHL": {
+      |			    "taxableIncome": 1011,
+      |			    "finalised": false
+      |	  	},
+      |		  "ukPropertyNonFHL": {
+      |		    	"taxableIncome": 1011,
+      |			    "finalised": false
+      |		  },
+      |		  "ukSavings": [{
+      |			    "savingsAccountId": "AA123456789",
```
Comment (jeremystone) [inline test/v1/models/response/getCalculation/GetCalculationResponseSpec.scala]: Different values for taxableIncome and/or savingsAccountName

File: test/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetailSpec.scala [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      "return an empty EoyEstimateDetail object" in {
+        eoyEstimateDetailDesJsonAllWrongIncomeSourceTypes.as[EoyEstimateDetail].isEmpty shouldBe true
+      }
+    }
+
+    /*
+    //TODO: Implement behaviour change for filter to surface read errors
+    "read from invalid Json" should {
+      "return a JsError" in {
+        eoyEstimateDetailInvalidJson.validate[EoyEstimateDetail] shouldBe false
```
Comment (jeremystone) [inline test/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetailSpec.scala]:
```suggestion
        eoyEstimateDetailInvalidJson.validate[EoyEstimateDetail] shouldBe a[JsError]
```

File: test/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetailSpec.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    "read from Json where all objects have an invalid income source type" should {
+      "return an empty EoyEstimateDetail object" in {
+        eoyEstimateDetailDesJsonAllWrongIncomeSourceTypes.as[EoyEstimateDetail].isEmpty shouldBe true
+      }
+    }
+
+    /*
+    //TODO: Implement behaviour change for filter to surface read errors
```
Comment (jeremystone) [inline test/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetailSpec.scala]: This fails. Implementing the change to the Reads suggested above (removing unnecessary asOpt usage) may fix it.

File: test/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetailSpec.scala [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        eoyEstimateDetailDesJson.validate[EoyEstimateDetail] shouldBe a[JsSuccess[_]]
+      }
+      "containing the expected EoyEstimateDetail object" in {
+        eoyEstimateDetailDesJson.as[EoyEstimateDetail] shouldBe eoyEstimateDetailResponse
+      }
+    }
+
+    "read from Json with missing optional fields" should {
+      "return the expected EoyEstimateDetail object" in {
+        eoyEstimateDetailDesJsonMissingOptionals.as[EoyEstimateDetail] shouldBe eoyEstimateDetailResponseReduced
```
Comment (jeremystone) [inline test/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetailSpec.scala]: This still has self-employment and savings elements. 

Are you testing that these are None in the written response when empty?

File: app/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateSelfEmployment.scala [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.models.response.getCalculation.endOfYearEstimate.detail
+
+import play.api.libs.functional.syntax._
+import play.api.libs.json.{ JsPath, Json, OWrites, Reads }
+
+case class EoyEstimateSelfEmployment(selfEmploymentID: String, taxableIncome: BigInt, finalised: Option[Boolean])
```
[the remark this answers is not in the document]
Comment (BetaDraconis) [reply app/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateSelfEmployment.scala]: changed to match spec

File: test/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetailSpec.scala (hunk 6 above)
Comment (jeremystone) [reply test/v1/models/response/getCalculation/endOfYearEstimate/detail/EoyEstimateDetailSpec.scala]: ☝️ just this one todo maybe? You check for reading empty Json but not reading Json with no savings or self-employment elements.

**Pull request merged**